File: Java/src/Shape/Shape3D/Shape3D.java
package Shape.Shape3D;

import Color.ColorRGB;
import Interface.ICalculable;
import Shape.Point2D;
import Shape.Point3D;
import Shape.Shape;
import Shape.Shape2D.Shape2D;

import java.util.ArrayList;

public abstract class Shape3D extends Shape implements ICalculable<Shape3D> {

    protected ArrayList<Point3D> points;
    protected Shape2D baseShape;

    public Shape3D(ArrayList<Point3D> p, Shape2D bShape, String name) {
        super(name);
        baseShape = bShape;
        points = p;
        adjustPoints(baseShape.getPoints());
    }

    public Shape3D(ArrayList<Point3D> p, Shape2D bShape, String name, ColorRGB color) {
        super(name, color);
        baseShape = bShape;
        points = p;
        adjustPoints(baseShape.getPoints());
    }

    public Shape3D(Shape3D s3d) {
        super(s3d);
        points = s3d.points;
        baseShape = s3d.baseShape;
    }

    public abstract double volume();

    public abstract double totalSurface();

    public int getNumberVertex() {
        return points.size();
    }

    public ArrayList<Point3D> getPoints() {
        return points;
    }

    public Shape2D getBaseShape() {
        return baseShape;
    }

    private void adjustPoints(ArrayList<Point2D> bs_points) {
        if (points == null)
            return;
        for (Point2D point : bs_points) {
            Point3D tmp = Point3D.p2dTop3d(point);
            boolean found = false;
            for (Point3D p3d : points) {
                if (tmp == p3d) {
                    found = true;
                }
            }
            if (!found) {
                points.add(tmp);
            }
        }
    }

    public boolean equalsPoints(ArrayList<Point3D> p) {
        if (points == null || p == null)
            return false;
        if (points.size() != p.size())
            return false;
        for (Point3D points : points) {
            if (!p.contains(points)) {
                return false;
            }
        }
        return true;
    }

    public boolean equals(Object obj) {
        if (super.equals(obj)) {
            Shape3D s3d = (Shape3D) obj;
            return equalsPoints(s3d.points);
        }
        return false;
    }
}

File: cpp/KalkCpp/Shape/Shape3D/shape3d.cpp
#include "shape3d.h"
#include <iostream>



Shape3D::Shape3D(const std::vector<Point3D>& vp3d, Shape2D* bs, std::string name, ColorRGB* color): Shape(name, color), points(vp3d), baseShape(bs)
{
    adjustPoints(bs->getPoints());
}



Shape3D::Shape3D(const Shape3D& s3d): Shape(s3d), points(s3d.points), baseShape(s3d.baseShape->clone())
{

}


bool Shape3D::operator ==(const Shape& s) const
{
    if(Shape::operator ==(s))
    {
        const Shape3D& s3d = static_cast<const Shape3D&>(s);
        if(points == s3d.points && *baseShape == *(s3d.baseShape))
        {
            return true;
        }
    }
    return false;
}

Shape3D& Shape3D::operator=(const Shape3D& s3d)
{
    if(this != &s3d){
        Shape::operator =(s3d);
        points = s3d.points;
        delete baseShape;
        baseShape = s3d.baseShape->clone();
    }
    return *this;
}



unsigned int Shape3D::getNumberVertex() const
{
    return points.size();
}



const std::vector<Point3D>& Shape3D::getPoints() const
{
    return points;
}



void Shape3D::adjustPoints(const std::vector<Point2D>& bs_points)
{
	for(std::vector<Point2D>::const_iterator it = bs_points.begin(); it != bs_points.end(); ++it)
	{
		Point3D tmp = Point3D::p2dTop3d(*it);
		bool found = false;
		for(std::vector<Point3D>::const_iterator it_t = points.begin(); it_t != points.end() && !found; ++it_t)
		{
			if(tmp == *it_t){
				found = true;
			}
		}
		if(!found){
			points.push_back(tmp);
		}
	}
}



const Shape2D* Shape3D::getBaseShape() const
{
    return baseShape;
}



Shape3D::~Shape3D()
{
    delete baseShape;
}
